rust/best-effort-response-calls/src/berc_backend/src/lib.rs

```diff
@@ -0,0 +1,106 @@
+use ic_cdk::call::CallError;
+use ic_cdk::prelude::*;
+
+/// Demonstrates that the system accepts best-effort calls, and that the receiver observes a deadline.
+///
+/// Invokes another method either with a best-effort response or a guaranteed response call,
+/// depending on the value of `use_best_effort_response`. Returns the deadline as observed by the
+/// receiver method.
+#[ic_cdk::update]
+async fn demonstrate_deadlines(use_best_effort_response: bool) -> u64 {
+    let call_builder = Call::new(ic_cdk::id(), "deadline_in_update").with_args(());
+    let call_builder = if use_best_effort_response {
+        call_builder.change_timeout(1)
+    } else {
+        call_builder.with_guaranteed_response()
+    };
+    let res: (u64,) = call_builder
+        .call()
+        .await
+        .expect("Didn't expect the call to fail");
+    res.0
+}
+
+/// Endpoint that demonstrates that timeouts can trigger with best-effort responses.
+/// It calls busy() with a timeout of 1 second, which is not enough to complete the
+/// execution, triggering a SYS_UNKNOWN error on the call. We return a bool to
+/// the caller to indicate whether a timeout has occured (true = yes, false = no)
+#[ic_cdk::update]
+async fn demonstrate_timeouts() -> bool {
+    let res: CallResult<(u64,)> = Call::new(ic_cdk::id(), "busy")
+        .change_timeout(1)
+        .with_args(((),))
+        .call()
+        .await;
+    match res {
+        Err(CallError::CallRejected(RejectCode::SysUnknown, s)) => {
+            ic_cdk::println!("SysUnknown: {:?}", s);
+            true
+        }
+        Err(e) => {
+            ic_cdk::println!("Unexpected error returned by the call: {:?}", e);
+            false
+        }
+        Ok((r, )) => {
+            ic_cdk::println!("Unexpected successful result returned by the call: {:?}", r);
+            false
+        }
+    }
+}
+
+/// Busy endpoint that just wastes a lot of instruction to trigger multi-round
+/// execution, which in turn can trigger timeouts on best-effort response calls
+/// to this endpoint
+#[ic_cdk::update]
+async fn busy() -> u64 {
+    const ROUNDS: u32 = 5;
+    const INSTRUCTIONS_PER_SLICE: u32 = 2_000_000_000;
+    const TOTAL_INSTRUCTIONS: u64 = ROUNDS as u64 * INSTRUCTIONS_PER_SLICE as u64;
+
+    while ic_cdk::api::performance_counter(0) < TOTAL_INSTRUCTIONS {}
+    ic_cdk::api::performance_counter(0)
+}
+
+#[ic_cdk::update]
+async fn deadline_in_update() -> u64 {
+    ic_cdk::api::call::msg_deadline()
+}
+
+#[ic_cdk::query]
+async fn deadline_in_query() -> u64 {
+    ic_cdk::api::call::msg_deadline()
+}
+
+#[ic_cdk::query(composite = true)]
+async fn deadline_in_composite_query() -> u64 {
+    ic_cdk::api::call::msg_deadline()
+}
+
+#[ic_cdk::query(composite = true)]
+async fn test_deadlines_in_composite_query() -> (u64, u64) {
```

Comment:
Do we actually do anything at all with deadlines in composite queries, beyond exposing them to the callee? Should we?

Reply:
We ended up [specifying this](https://github.com/dfinity/interface-spec/pull/268/files#diff-b9a833d6114611a080aee4cd57618feb45aa044bcc11105d44b155b5bbde00ebR1563):
> For other calls (ingress messages and all calls to query and composite query methods, including calls in replicated mode) a deadline of 0 is returned.

The implementation is consistent with that. I've added this to the README now.